src/xrpld/app/tx/detail/AMMClawback.cpp

```diff
@@ -0,0 +1,283 @@
+//------------------------------------------------------------------------------
+/*
+  This file is part of rippled: https://github.com/ripple/rippled
+  Copyright (c) 2024 Ripple Labs Inc.
+
+  Permission to use, copy, modify, and/or distribute this software for any
+  purpose  with  or without fee is hereby granted, provided that the above
+  copyright notice and this permission notice appear in all copies.
+
+  THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+  WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+  MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+  ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+  WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+  ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/misc/AMMHelpers.h>
+#include <xrpld/app/misc/AMMUtils.h>
+#include <xrpld/app/tx/detail/AMMClawback.h>
+#include <xrpld/app/tx/detail/AMMWithdraw.h>
+#include <xrpld/ledger/Sandbox.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/protocol/AMMCore.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/Protocol.h>
+#include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/st.h>
+#include <tuple>
+
+namespace ripple {
+
+NotTEC
+AMMClawback::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureAMMClawback))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;  // LCOV_EXCL_LINE
+
+    if (ctx.tx.getFlags() & tfAMMClawbackMask)
+        return temINVALID_FLAG;
+
+    AccountID const issuer = ctx.tx[sfAccount];
+    AccountID const holder = ctx.tx[sfHolder];
+
+    if (issuer == holder)
+    {
+        JLOG(ctx.j.trace())
+            << "AMMClawback: holder cannot be the same as issuer.";
+        return temMALFORMED;
+    }
+
+    std::optional<STAmount> const clawAmount = ctx.tx[~sfAmount];
+    auto const asset = ctx.tx[sfAsset];
+
+    if (isXRP(asset))
+        return temMALFORMED;
+
+    if (asset.account != issuer)
+    {
+        JLOG(ctx.j.trace()) << "AMMClawback: Asset's account does not "
+                               "match Account field.";
+        return temBAD_ASSET_ISSUER;
```

Comment:
nit (Optional): It would be great if we could make these agnostic. I use them in another PR, and other PR's to this point have used temMALFORMED/BAD_AMOUNT. I think its a valuable addition. In general the error is (the tx issuer does not match the ledger entry issuer).

Reply:
I was mistaken on this. This is not checking the ledger index. Because its not, and this is technically just a MALFORMED txn I would suggest we use malformed here. We dont have an infinite number of error codes and so creating 2 new error codes just for this where we have used MALFORMED in the past isnt a good practice imo.

Reply:
I can change both of the `temBAD_ASSET_AMOUNT`  and `temBAD_ASSET_ISSUER` to `temMALFORMED`. In the beginning, I wanted to provide the accurate error message to the client so that the client knows what was wrong with the request.

Reply:
updated in the recent commit to `temMALFORMED`